core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,377 @@
+      i++;
+    }
+    X509ExtendedTrustManager newDelegateManager = createDelegateTrustManager(keyStore);
+    this.delegateManager = newDelegateManager;
+  }
+
+  private static X509ExtendedTrustManager createDelegateTrustManager(KeyStore keyStore)
+      throws CertificateException, KeyStoreException, NoSuchAlgorithmException {
+    TrustManagerFactory tmf = TrustManagerFactory.getInstance(
+        TrustManagerFactory.getDefaultAlgorithm());
+    tmf.init(keyStore);
+    X509ExtendedTrustManager delegateManager = null;
+    TrustManager[] tms = tmf.getTrustManagers();
+    // Iterate over the returned trust managers, looking for an instance of X509TrustManager.
+    // If found, use that as the delegate trust manager.
+    for (int j = 0; j < tms.length; j++) {
+      if (tms[j] instanceof X509ExtendedTrustManager) {
+        delegateManager = (X509ExtendedTrustManager) tms[j];
+        break;
+      }
+    }
+    if (delegateManager == null) {
+      throw new CertificateException(
+          "Instance delegateX509TrustManager is null. Failed to initialize");
+    }
+    return delegateManager;
+  }
+
+  private void checkTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine,
+      Socket socket, boolean checkingServer) throws CertificateException {
+    if (this.verification != Verification.InsecurelySkipAllVerification) {
+      if (chain == null || chain.length == 0) {
+        throw new CertificateException(
+            "Want certificate verification but got null or empty certificates");
+      }
+      if (this.delegateManager == null) {
+        throw new CertificateException("Credential hasn't been provided yet");
+      }
+      if (checkingServer) {
+        if (this.verification == Verification.CertificateAndHostNameVerification
+            && sslEngine == null) {
+          throw new CertificateException(
+              "SSLEngine is null. Couldn't check host name");
+        }
+        String algorithm = this.verification == Verification.CertificateAndHostNameVerification
+            ? "HTTPS" : "";
+        SSLParameters sslParams = sslEngine.getSSLParameters();
+        sslParams.setEndpointIdentificationAlgorithm(algorithm);
+        sslEngine.setSSLParameters(sslParams);
+        delegateManager.checkServerTrusted(chain, authType, sslEngine);
+      } else {
+        delegateManager.checkClientTrusted(chain, authType, sslEngine);
+      }
+    }
+    // Perform the additional peer cert check.
+    if (sslEngine != null && socketAndEnginePeerVerifier != null) {
+      socketAndEnginePeerVerifier.verifyPeerCertificate(chain, authType, sslEngine);
+    } else if (socket != null && socketAndEnginePeerVerifier != null) {
+      socketAndEnginePeerVerifier.verifyPeerCertificate(chain, authType, socket);
+    } else if (peerVerifier != null) {
+      peerVerifier.verifyPeerCertificate(chain, authType);
+    } else {
+      log.log(Level.INFO, "No peer verifier is specified");
```

Comment:
Why is this being logged at a default-logged level? There doesn't seem to be any issue here and nothing to correct. And this will log for every incoming connection. I don't see the value of this log so would suggest removing it, but at the very least make it FINE log level.